### PR TITLE
Updating the VAE tutorial

### PR DESCRIPTION
**Pull request by neerajprad**
Changes:
 - I have updated the code in the code-blocks and added it as runnable code cells. Also added a smoke test for the tutorial.
 - Updated the test elbo loss figure which was not correct earlier.
 - Minor cosmetic changes.

PS: I strongly suggest that we convert our other tutorials like ss-vae to have runnable code cells. It is a bit of a time investment, but will make it easier to maintain these tutorials over time.

### CODE REVIEW

**Comment by neerajprad** [comment]
> are the examples up to date?

The example is other wise up to date.

> do we need all that cuda logic in the tutorials? since presumably users will use the actual example if running on gpu

I ran the tutorial on the gpu. It takes quite a long time otherwise to even run 10 epochs. So I suppose we should have the cuda flag so that people can toy around with it.

**Comment by martinjankowiak** [comment]
@karalets looks like this vanilla VAE is getting a test elbo of about -102 with a 50 dimensional latent space and 400 hidden units in the nn... how does that seem?

File: tutorial/source/vae.ipynb [hunk 1]
```diff
@@ -85,13 +85,47 @@
...
+    "from pyro.optim import Adam"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pyro.enable_validation(False)\n",
```
Comment (fritzo) [inline tutorial/source/vae.ipynb]: Should we set `pyro.enable_validation(smoke_test)` to be safe?

Comment (fritzo) [reply tutorial/source/vae.ipynb]: Ok. This would be a nice use case for #988 
```py
pyro.sample("obs", dist.Bernoulli(prediction), obs=dist.Bernoulli(ground_truth))
```
or even for parallel enumeration (if it doesn't add too much overhead):
```py
truth = pyro.sample("truth", dist.Bernoulli(ground_truth),
                    infer={"enumerate": "parallel"})
pyro.sample("obs", dist.Bernoulli(prediction), obs=truth)
```
But we can address this later...

File: tutorial/source/vae.ipynb [hunk 2]
```diff
@@ -106,13 +140,24 @@
...
-    "```python\n",
+    "Next we define a PyTorch module that encapsulates our decoder network:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
[the remark this answers is not in the document]
Comment (martinjankowiak) [reply tutorial/source/vae.ipynb]: setup the non-linearity => setup the non-linearities

[the remark this answers is not in the document]
Comment (martinjankowiak) [reply tutorial/source/vae.ipynb]: if the fudge disappeared kill this comment: fixing numerical instabilities of sigmoid with a fudge

File: tutorial/source/vae.ipynb [hunk 3]
```diff
@@ -161,60 +215,91 @@
...
-    "    # (value will be sampled by guide when computing the ELBO)\n",
-    "    z = pyro.sample(\"latent\", dist.Normal(z_loc, z_scale))\n",
-    "\n",
-    "    # decode the latent code z\n",
-    "    mu_img = self.decoder(z)\n",
-    "    # score against actual images\n",
-    "    pyro.sample(\"obs\", dist.Bernoulli(x.reshape(-1, 784)), obs=mu_img)\n",
-    "```\n",
-    "\n",
+    "    with pyro.iarange(\"data\", x.size(0)):\n",
```
Comment (martinjankowiak) [inline tutorial/source/vae.ipynb]: this iarange needs explaining in the text

File: tutorial/source/vae.ipynb [hunk 4]
```diff
@@ -161,60 +215,91 @@
...
+    "        # setup hyperparameters for prior p(z)\n",
+    "        z_loc = x.new_zeros(torch.Size((x.size(0), self.z_dim)))\n",
+    "        z_scale = x.new_ones(torch.Size((x.size(0), self.z_dim)))\n",
+    "        # sample from prior (value will be sampled by guide when computing the ELBO)\n",
+    "        z = pyro.sample(\"latent\", dist.Normal(z_loc, z_scale).independent(1))\n",
+    "        # decode the latent code z\n",
+    "        loc_img = self.decoder.forward(z)\n",
+    "        # score against actual images\n",
+    "        pyro.sample(\"obs\", dist.Bernoulli(loc_img).independent(1), obs=x.reshape(-1, 784))"
```
Comment (martinjankowiak) [inline tutorial/source/vae.ipynb]: same with `.independent(1)`

File: tutorial/source/vae.ipynb [hunk 5]
```diff
@@ -161,60 +215,91 @@
...
+    "def guide(self, x):\n",
     "    # register PyTorch module `encoder` with Pyro\n",
     "    pyro.module(\"encoder\", self.encoder)\n",
-    "    # use the encoder to get the parameters used to define q(z|x)\n",
-    "    z_loc, z_scale = self.encoder(x)\n",
-    "    # sample the latent code z\n",
-    "    pyro.sample(\"latent\", dist.Normal(z_loc, z_scale))\n",
-    "```\n",
-    "\n",
+    "    with pyro.iarange(\"data\", x.size(0)):\n",
```
Comment (martinjankowiak) [inline tutorial/source/vae.ipynb]: worth pointing to iarange/ind in guide too

File: tutorial/source/vae.ipynb [hunk 6]
```diff
@@ -247,81 +365,182 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run options\n",
+    "LEARNING_RATE = 1.0e-3\n",
+    "USE_CUDA = False\n",
+    "NUM_EPOCHS = 1 if smoke_test else 100\n",
```
Comment (martinjankowiak) [inline tutorial/source/vae.ipynb]: explain smoke_test with short comment

**Comment by neerajprad** [comment]
Thanks for reviewing. I will address all of the comments shortly.

>  looks like this vanilla VAE is getting a test elbo of about -102 with a 50 dimensional latent space and 400 hidden units in the nn... how does that seem?

What would have been your expectation otherwise? For context, we had it go up to `-80` earlier with the same model, because Bernoulli was not using binary cross entropy for scoring.

**Comment by neerajprad** [comment]
@martinjankowiak - Let me know if you have any more comments.

File: examples/vae/vae.py [hunk 7]
```diff
@@ -46,10 +46,10 @@ def forward(self, x):
 class Decoder(nn.Module):
     def __init__(self, z_dim, hidden_dim):
         super(Decoder, self).__init__()
-        # setup the three linear transformations used
+        # setup the two linear transformations used
         self.fc1 = nn.Linear(z_dim, hidden_dim)
         self.fc21 = nn.Linear(hidden_dim, 784)
-        # setup the non-linearity
+        # setup the non-linearities
```
Comment (martinjankowiak) [inline examples/vae/vae.py]: looks good but you have `fudge` in the decoder in vae.py but not in the notebook?

Comment (neerajprad) [reply examples/vae/vae.py]: Good point; it's not needed. Removed.

Comment (martinjankowiak) [reply examples/vae/vae.py]: but can you confirm you can get good numbers (no crashing etc.) without the fudge?

Comment (neerajprad) [reply examples/vae/vae.py]: The test graph was generated from the notebook itself, so I can confirm that it doesn't crash, and gives reasonable results. Let me generate some sample images too, and validate.
